### DISCUSSION TITLE
omit empty strings fields mongoose

### DISCUSSION THREAD

**Discussion by samislam** [Automattic/mongoose#12226] [Q&A]
I have the following schema:
```js
const mySchema = new mongoose.Schema({
   x: String,
   y: String
})
```

when a user from the front-end requests in his body:
```js
req.body = {
   'x' : '',
   'y': ''
}
```
this results in creating a field in MongoDB, but with an empty string.
I need a way to prevent this behavior by setting the empty strings to be undefined somehow. 

Is there such an option in Mongoose? or do I have to predict my own middlewares for that?

**Comment by vkarpov15** [comment]
Custom validators:

```javascript
const mySchema = new mongoose.Schema({
   x: { type: String, validate: v == null || v.length > 0 },
   y: { type: String, validate: v == null || v.length > 0 }
})
```

**Reply by samislam** [reply]
Imagine having to do that everywhere

**Reply by Uzlopak** [reply]
encapsulate in a function and reuse it.

**Reply by vkarpov15** [reply]
Or just set a global default validator as follows. In general, you can set defaults per schema type (e.g. trim all strings by default or validate that all numbers are positive by default) [globally using `SchemaType.set()`](https://thecodebarbarian.com/whats-new-in-mongoose-54-global-schematype-configuration.html).

```
const mongoose = require('mongoose');

// Set a default validator on all string properties
mongoose.Schema.Types.String.set('validate', v => v == null || v > 0);

const mySchema = new mongoose.Schema({
   x: { type: String },
   y: { type: String }
});

const Test = mongoose.model('Test', mySchema);

const doc = new Test({ x: '' });
// Error: Test validation failed: x: Validator failed for path `x` with value ``
console.log(doc.validateSync());
```

**Reply by samislam** [reply]
@vkarpov15 , I believe we should consider making it the default behavior, _or partly_, it's good for the front-end developer to think that _"Okay, the `email` field should be a string so I'll call the `.slice(0, 3)` method to do xyz!"_, but on the other hand, it's a pain for the back-end developer when it comes to many other things, for instance, think of the following fields:
```js
email: {
  type: String,
  unique: true,
  sparse: true
}
```
if you have these properties enabled in your mongoose schema on the `email` field, you'll face the problem of when the consumer of your API tries to create two users without specifying an email by sending the request as follows in the request body:
```js
email: ""
```
In this case, the `sparse` prop will do nothing because it will think that email ("empty string") and email ("empty string") are not `null`, and hence the `unique` property is going to throw a **E11000 duplicate key error collection mongoDB error**.

maybe either we should make the default behavior to convert the empty strings to `undefined`/`null` internally automatically, or at least we should do that when we specify the `sparse` and `unique` properties at the same time.

I also found this problem causing some noise when I was using the [validator](https://www.npmjs.com/package/validator) package.. I was validating if one of the input fields is a valid phone number, and I still remember I didn't sleep at that night until somehow I discovered that [validator](https://www.npmjs.com/package/validator) was throwing a validation error because the empty string (`""`) was not the valid phone number.

**Reply by samislam** [reply]
> Custom validators:
> 
> ```js
> const mySchema = new mongoose.Schema({
>    x: { type: String, validate: v == null || v.length > 0 },
>    y: { type: String, validate: v == null || v.length > 0 }
> })
> ```

By the way, even though this works OK for the validations, this keeps writing empty strings to the database. which causes problems when using both of `sparse` and `unique` properties on one field as I mentioned above.

**Reply by vkarpov15** [reply]
I disagree about treating empty strings as invalid default behavior. It is easy enough to work around, especially given that `required` also works:

```javascript
const mongoose = require('mongoose');

// Set a default validator on all string properties
//mongoose.Schema.Types.String.set('validate', v => v == null || v > 0);

const mySchema = new mongoose.Schema({
   x: { type: String, required: true },
   y: { type: String, required: true }
});

const Test = mongoose.model('Test', mySchema);

const doc = new Test({ x: '' });
// Error: Test validation failed: x: Validator failed for path `x` with value ``
console.log(doc.validateSync());

```
